File: src/main/java/io/github/jolkert/discordbotlink/jda/DiscordBot.java
package io.github.jolkert.discordbotlink.jda;

import io.github.jolkert.discordbotlink.DiscordBotLink_LEGACY;
import io.github.jolkert.discordbotlink.data.UserDataHolder;
import io.github.jolkert.discordbotlink.jda.config.BotConfig;
import io.github.jolkert.discordbotlink.jda.listener.CommandHandler;
import io.github.jolkert.discordbotlink.jda.listener.LinkChannelListener;
import io.github.jolkert.discordbotlink.jda.listener.UserUpdateListener;
import net.dv8tion.jda.api.JDA;
import net.dv8tion.jda.api.JDABuilder;
import net.dv8tion.jda.api.entities.*;
import net.dv8tion.jda.api.entities.channel.concrete.TextChannel;
import net.dv8tion.jda.api.events.session.ReadyEvent;
import net.dv8tion.jda.api.hooks.ListenerAdapter;
import net.dv8tion.jda.api.interactions.commands.OptionType;
import net.dv8tion.jda.api.interactions.commands.build.Commands;
import net.dv8tion.jda.api.requests.GatewayIntent;
import net.dv8tion.jda.api.utils.MemberCachePolicy;
import net.dv8tion.jda.api.utils.concurrent.Task;
import net.fabricmc.loader.api.FabricLoader;

import javax.security.auth.login.LoginException;
import java.util.List;

public class DiscordBot extends ListenerAdapter
{
	private JDA jda;
	private final UserDataHolder userData;
	public Guild primaryGuild;
	private final BotConfig config;

	private TextChannel linkChannel = null;

	private static final String FILE_PATH = FabricLoader.getInstance().getConfigDir().toString() + "/DiscordBotLink/user_data.json";
	
	
	public DiscordBot(BotConfig config) throws LoginException
	{
		jda = JDABuilder.createLight(config.token(), GatewayIntent.getIntents(GatewayIntent.ALL_INTENTS))
				.addEventListeners(new CommandHandler(config.prefix()), this, new UserUpdateListener(), new LinkChannelListener())
				.setActivity(Activity.playing("Prefix: " + config.prefix()))
				.setMemberCachePolicy(MemberCachePolicy.ALL)
				.build();

		jda.updateCommands().addCommands(
				Commands.slash("whitelist", "Adds a user to the Minecraft whitelist")
						.addOption(OptionType.STRING, "username", "Your minecraft username. You can only add one Minecraft account per Discord account.")
		).queue();

		userData = new UserDataHolder(FILE_PATH);
		this.config = config;
	}
	
	public UserDataHolder getUserData()
	{
		return userData;
	}
	
	@Override
	public void onReady(ReadyEvent event)
	{
		jda = event.getJDA();
		
		primaryGuild = jda.getGuilds().stream().findFirst().orElseThrow();
		
		Task<List<Member>> _ignored = primaryGuild.loadMembers();
		
		DiscordBotLink_LEGACY.Logger.info("Discord bot ready!");
	}

	public BotConfig getConfig()
	{
		return config;
	}

	public TextChannel getLinkChannel()
	{
		if (linkChannel == null)
			linkChannel = jda.getTextChannelById(config.linkChannelId());

		return linkChannel;
	}

	public JDA getJda()
	{
		return jda;
	}
}

File: src/main/kotlin/io/github/jolkert/discordbotlink/jda/DiscordBot.kt
package io.github.jolkert.discordbotlink.jda

import io.github.jolkert.discordbotlink.DiscordBotLink
import io.github.jolkert.discordbotlink.DiscordBotLink.CONFIG_FILE
import io.github.jolkert.discordbotlink.DiscordBotLink.LOGGER
import io.github.jolkert.discordbotlink.extension.pascalToSnake
import io.github.jolkert.discordbotlink.extension.pascalToTitle
import io.github.jolkert.discordbotlink.jda.command.WhitelistButtonCommand
import io.github.jolkert.discordbotlink.jda.data.BotConfig
import io.github.jolkert.discordbotlink.jda.listener.CommandHandler
import io.github.jolkert.discordbotlink.jda.listener.LinkChannelListener
import io.github.jolkert.discordbotlink.jda.listener.UserUpdateListener
import net.dv8tion.jda.api.JDABuilder
import net.dv8tion.jda.api.entities.Activity
import net.dv8tion.jda.api.entities.Guild
import net.dv8tion.jda.api.entities.User
import net.dv8tion.jda.api.entities.channel.concrete.TextChannel
import net.dv8tion.jda.api.events.session.ReadyEvent
import net.dv8tion.jda.api.hooks.ListenerAdapter
import net.dv8tion.jda.api.requests.GatewayIntent
import net.dv8tion.jda.api.utils.MemberCachePolicy
import net.minecraft.text.MutableText
import net.minecraft.text.Text

class DiscordBot(private val config: BotConfig) : ListenerAdapter()
{
	private val jda = JDABuilder.createLight(config.token, GatewayIntent.getIntents(GatewayIntent.ALL_INTENTS))
		.addEventListeners(this, CommandHandler(config.commandPrefix), UserUpdateListener, LinkChannelListener, WhitelistButtonCommand.Companion.ButtonListener)
		.setActivity(Activity.playing("Prefix: ${config.commandPrefix}"))
		.setMemberCachePolicy(MemberCachePolicy.ALL)
		.build()
		.apply { updateCommands().addCommands(CommandHandler.commands.map { it.slashCommandData }).queue() }

	private lateinit var primaryGuild: Guild
	var linkChannel: TextChannel? = null
		private set

	val self: User
		get() = jda.selfUser

	var kickMessage: MutableText = if (config.kickMessage.isNotEmpty()) Text.literal(config.kickMessage) else Text.translatable("multiplayer.disconnect.not_whitelisted")
		set(value)
		{
			field = value
			config.toProperties().apply {
				setProperty(BotConfig::kickMessage.name.pascalToSnake(), value.string)
				CONFIG_FILE.printWriter().also {
					store(it, "Config for ${this::class.simpleName!!.pascalToTitle()}")
				}.close()
			}
		}

	override fun onReady(event: ReadyEvent)
	{
		primaryGuild = jda.guilds.first()
		primaryGuild.loadMembers()
		linkChannel = jda.getTextChannelById(config.linkChannelId)

		LOGGER.info("Discord bot ready!")
		LOGGER.info("Primary Guild: ${primaryGuild.name}")
	}

	fun sendMessageInLinkChannel(message: String)
	{
		linkChannel?.sendMessage(message)?.queue()
	}
}